nixops/backends/libvirtd.py

```diff
@@ -130,37 +173,79 @@ def create(self, defn, check, allow_reboot, allow_recreate):
                 self.log('Failed to register domain XML with the hypervisor')
                 return False
 
-            newEnv = copy.deepcopy(os.environ)
-            newEnv["NIXOPS_LIBVIRTD_PUBKEY"] = self.client_public_key
-            base_image = self._logged_exec(
-                ["nix-build"] + self.depl._eval_flags(self.depl.nix_exprs) +
-                ["--arg", "checkConfigurationOptions", "false",
-                 "-A", "nodes.{0}.config.deployment.libvirtd.baseImage".format(self.name),
-                 "-o", "{0}/libvirtd-image-{1}".format(self.depl.tempdir, self.name)],
-                capture_stdout=True, env=newEnv).rstrip()
-
-            if not os.access(defn.image_dir, os.W_OK):
-                raise Exception('{} is not writable by this user or it does not exist'.format(defn.image_dir))
-
-            self.disk_path = self._disk_path(defn)
-            shutil.copyfile(base_image + "/disk.qcow2", self.disk_path)
-            # Rebase onto empty backing file to prevent breaking the disk image
-            # when the backing file gets garbage collected.
-            self._logged_exec(["qemu-img", "rebase", "-f", "qcow2", "-b",
-                               "", self.disk_path])
-            os.chmod(self.disk_path, 0660)
             self.vm_id = self._vm_id()
 
         self.start()
         return True
 
-    def _disk_path(self, defn):
-        return "{0}/{1}.img".format(defn.image_dir, self._vm_id())
+    def _prepare_storage_volume(self):
+        self.logger.log("preparing disk image...")
+        newEnv = copy.deepcopy(os.environ)
+        newEnv["NIXOPS_LIBVIRTD_PUBKEY"] = self.client_public_key
+
+        temp_image_path = os.path.join(self.depl.tempdir, 'libvirtd-image-{}'.format(self.name))
+        base_image = self._logged_exec(
+            ["nix-build"] + self.depl._eval_flags(self.depl.nix_exprs) +
+            ["--arg", "checkConfigurationOptions", "false",
+             "-A", "nodes.{0}.config.deployment.libvirtd.baseImage".format(self.name),
+             "-o", temp_image_path],
+            capture_stdout=True, env=newEnv).rstrip()
+
+        temp_disk_path = os.path.join(self.depl.tempdir, 'disk-{}.qcow2'.format(self.name))
+        shutil.copyfile(os.path.join(temp_image_path, 'disk.qcow2'), temp_disk_path)
+        # Rebase onto empty backing file to prevent breaking the disk image
+        # when the backing file gets garbage collected.
+        self._logged_exec(["qemu-img", "rebase", "-f", "qcow2", "-b",
```

Comment:
just wondering if the plain qemu-img could be better replaced with `_get_qemu_executable` ?

Reply:
ok cool I hope this gets merged soon